### ISSUE TITLE
A way to dynamically SkipOnInsert based on field value

### ISSUE THREAD

**Issue by arwyl** [linq2db/linq2db#981]
A very questionable feature, but maybe it can be achieved.
I have a table with uniqueidentifier "Guid" field with default value (newsequentialid()).
Usually this field should be generated by DB, but in some cases (cross db sync) I need to set it from code. Is there a way to achieve this? Maybe something like custom attribute to set SkipOnInsert based on the field value?

Test class
```cs
		public class Test
		{
			[Identity, PrimaryKey, Column]
			public int Id { get; set; }

			[Column]
			public string Name { get; set; }

                        // SHOULD be in insert query only when equals Guid.Empty
			[Column("Guid")]
			public Guid Guid { get; set; }
		}
```

Create table query
```sql
CREATE TABLE [dbo].[Tests](
	[Id] [int] IDENTITY(1,1) NOT NULL,
	[Name] [varchar](140) NULL,
	[GUID] [uniqueidentifier] NOT NULL
 CONSTRAINT [PK___Tests] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80) ON [PRIMARY]
) ON [PRIMARY]
go;
ALTER TABLE [dbo].[Tests] ADD  DEFAULT (newsequentialid()) FOR [GUID]
```

**Comment by sdanyliv** [comment]
You can create MappingSchema and set attribute by itself. DataConnection should be crated with this schema.
A lot of samples here:
https://github.com/linq2db/linq2db/blob/master/Tests/Linq/Mapping/MappingSchemaTests.cs

**Comment by arwyl** [comment]
thanks a lot! I'll try that!

**Comment by sdanyliv** [comment]
Also there another Insert method with lambda parameter - here we ignore these attributes, you can control which fileds needs to be inserted.
If your sync databases are located on the same server there also much effective to use Insert from.

```cs
db.SomeItems.DatabaseName(“SourceDb”)
    .Where(filter)
    .Insert(db.SomeItems.DatabaseName(“DestDb”),
        i => new SomeItem 
        {
            Filed1 = i.Field1,
         ...
        });
```

**Comment by sdanyliv** [comment]
For bulk inserting Items, use db.BulkCopy method.

**Comment by arwyl** [comment]
I managed to find MappingSchema FluentMappingBuilder to set needed attributes, but it seems, that it somehow overrides the behaviour globally. Even if I add it via db.AddMappingScheme.

Two different DataConnection objects. Both of them add new MappingScheme. One with property SkipOnInsert=true, another with SkipOnInsert=false. But with the second data connection property is also skipped.

```
using System;
using System.Linq;
using LinqToDB;
using LinqToDB.Data;
using LinqToDB.Mapping;
using NUnit.Framework;

namespace Tests
{
	[TestFixture]
	public class Linq2DbSequenceGuidTests
	{
		[Table("Tests")]
		public class Test
		{
			[Identity, PrimaryKey, Column]
			public int Id { get; set; }

			[Column]
			public string Name { get; set; }

			[Column("GUID")]
			public Guid Guid { get; set; }

			public const string CreateScript = @"
CREATE TABLE [dbo].[Tests](
	[Id] [int] IDENTITY(1,1) NOT NULL,
	[Name] [varchar](140) NULL,
	[GUID] [uniqueidentifier] NOT NULL
 CONSTRAINT [PK___Tests] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 80) ON [PRIMARY]
) ON [PRIMARY]
";

			public const string SetDefaultScript = @"
ALTER TABLE [dbo].[Tests] ADD  DEFAULT (newsequentialid()) FOR [GUID]
";
		}

		const string ConfigString = "UnitTestsConnectionString";

		[SetUp]
		public void SetUp()
		{
			using (var db = new DataConnection(ConfigString))
			{
				db.Execute(Test.CreateScript);
				db.Execute(Test.SetDefaultScript);
			}
		}

		[TearDown]
		public void TearDown()
		{
			using (var db = new DataConnection(ConfigString))
			{
				db.DropTable<Test>();
			}
		}

		[Test]
		public void Linq2DbSequenceGuidTests_Insert_With_Default()
		{
			DataConnection.TurnTraceSwitchOn(System.Diagnostics.TraceLevel.Info);
			DataConnection.OnTrace = trace =>
			{
				if (trace.TraceInfoStep == TraceInfoStep.AfterExecute)
				{
					Console.WriteLine(trace.SqlText);
				}
			};

			using (var db = new DataConnection(ConfigString))
			{
				var ms = new MappingSchema();
				var mb = ms.GetFluentMappingBuilder();
				mb.Entity<Test>().Property(x => x.Guid).HasSkipOnInsert(true).HasSkipOnUpdate(true);

				db.AddMappingSchema(ms);

				var idObj = db.InsertWithIdentity(new Test
				{
					Name = "123"
				});

				Console.WriteLine(idObj);

				int id = Convert.ToInt32(idObj);

				var fromDb = db.GetTable<Test>().FirstOrDefault(x => x.Id == id);

				Console.WriteLine(fromDb.Guid);

				Assert.AreNotEqual(Guid.Empty, fromDb.Guid);
			}

			using (var db = new DataConnection(ConfigString))
			{
				var ms = new MappingSchema();
				var mb = ms.GetFluentMappingBuilder();
				mb.Entity<Test>().Property(x => x.Guid).HasSkipOnInsert(false).HasSkipOnUpdate(false);

				db.AddMappingSchema(ms);

				var newGuid = Guid.NewGuid();

				var idObj = db.InsertWithIdentity(new Test
				{
					Name = "123",
					Guid = newGuid
				});

				Console.WriteLine(idObj);

				int id = Convert.ToInt32(idObj);

				var fromDb = db.GetTable<Test>().FirstOrDefault(x => x.Id == id);

				Console.WriteLine(fromDb.Guid);

                                // FAILS..
				Assert.AreEqual(newGuid, fromDb.Guid);
			}

			DataConnection.TurnTraceSwitchOn(System.Diagnostics.TraceLevel.Off);
		}

	}
}

```


Trace string:

```
DECLARE @Name NVarChar(3) -- String
SET     @Name = N'123'

INSERT INTO [Tests]
(
	[Name]
)
VALUES
(
	@Name
)

SELECT SCOPE_IDENTITY()

1

DECLARE @id1 Int -- Int32
SET     @id1 = 1

SELECT TOP (1)
	[t1].[Id],
	[t1].[Name],
	[t1].[GUID]
FROM
	[Tests] [t1]
WHERE
	[t1].[Id] = @id1

b445f028-e400-e811-8b41-001c4220a504

DECLARE @Name NVarChar(3) -- String
SET     @Name = N'123'

INSERT INTO [Tests]
(
	[Name]
)
VALUES
(
	@Name
)

SELECT SCOPE_IDENTITY()

2

DECLARE @id1 Int -- Int32
SET     @id1 = 2

SELECT TOP (1)
	[t1].[Id],
	[t1].[Name],
	[t1].[GUID]
FROM
	[Tests] [t1]
WHERE
	[t1].[Id] = @id1

b545f028-e400-e811-8b41-001c4220a504
```


Can you please tell me what I am doing wrong?

**Comment by arwyl** [comment]
I have a theory, that the problem is in the default name of the new MappingScheme
because changing it to 

```
var ms = new MappingSchema(Guid.NewGuid().ToString());
```
helped

**Comment by Franki1986** [comment]
Is there anything new here?

**Comment by sdanyliv** [comment]
@Franki1986, PR is appreciated. 

You can start here: https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Linq/QueryRunner.Insert.cs#L39
But dont't forget to disable cache. We will help in avoiding such issue during PR review.

**Comment by Franki1986** [comment]
Any way to get the values that will be inserted at this point?

**Comment by Franki1986** [comment]
Aded some changes to my fork:

[SkipValuesOnInsert](https://github.com/linq2db/linq2db/compare/v2.6.4...Franki1986:v2.6.4.1)

I thought it would be nice to do something like this:

```
[Column("Column1"), SkipValuesOnInsert(0, 1)]
public Int32 Column1 { get; set; }
```

**Comment by sdanyliv** [comment]
Thinking about this one. Why just not create new method `InsertSmart` in your code which dynamically work with `IUpdatable` and `Set`?

**Comment by Franki1986** [comment]
If I use an InsertSmart, I would have to update my Inserts. With a new attribute I could use the same syntax everywhere and only change the mapping config..

**Comment by sdanyliv** [comment]
@Franki1986, create WIP PR which i can comment. There are a lot of problems :)

**Comment by Franki1986** [comment]
:-) Yeah I knew there would be unknown problems...

**Comment by lobster2012-user** [comment]
@Franki1986 
Ha-ha.
I was going to push the request with the addition of [SkipOnInsertDefaultValue] after reading your issue
Just now I saw your comment about your PR.

**Comment by Franki1986** [comment]
Yes.. Monday morning I am back to my pc and will commit the changes sdanyliv requested..

**Comment by MaceWindu** [comment]
implemented in #1598
